docs/coding-guidelines.md

```diff
@@ -333,6 +333,55 @@ The person writing the code will write the doc comments. Public APIs only. No ne
 
 Note: Public means callable by a customer, so it includes protected APIs. However, some public APIs might still be "for internal use only" but need to be public for technical reasons. We will still have doc comments for these APIs but they will be documented as appropriate.
 
+- Do not include empty XML comments.
+  - summary element must not be empty.
+  - param element must not be empty.
+
+Correct:
+
+```cs
+        /// <summary>
+        /// Calculates all fruit types
+        /// </summary>
+        /// <returns>A set of fruit types contained in this basket.</returns>
+        public ISet<FruitType> GetAllFruitTypes()
+
+        /// <summary>
+        /// Determines whether the fruit has seeds.
+        /// </summary>
+        /// <param name="fruitType">A fruit type</param>
+        /// <returns> Whether the fruit has seeds. </returns>
+        public bool HasSeeds(FruitType fruitType)
+        
+        /// <summary>
+        /// Determines whether the fruit is multiannual
+        /// </summary>
+        /// <returns> Whether the fruit is multiannual. </returns>
+        public bool IsMultiannual(FruitType fruitType)
+```
+
+Incorrect:
+
+```cs
+        /// <summary>
+        /// </summary>
+        /// <returns> </returns>
+        public ISet<FruitType> GetAllFruitTypes()
+
+        /// <summary>
+        /// Determines whether the fruit has seeds.
+        /// </summary>
+        /// <param name="fruitType"></param>
+        /// <returns> </returns>
+        public bool HasSeeds(FruitType fruitType)
+        
+        /// <summary>
+        /// Determines whether the fruit is multiannual
+        /// </summary>
+        /// <returns> </returns>
+        public bool IsMultiannual(FruitType fruitType)
```

Comment:
You added this after my review. I'm not sure about this. Now the XML comment doesn't match the signature :\

Reply:
Yes, all. No, it's okay if you want to require the names be repeated in the description. Just never seen arbitrarily missing arguments before.

Reply:
I believe that the analyzers built into roslyn even warn when you have xmldoc that is missing for a parameter.

Reply:
> Yes, all. No, it's okay if you want to require the names be repeated in the description. Just never seen arbitrarily missing arguments before.

My suggestion is all or nothing. 
You either describe *all* parameters, or none. Same applies to *every* xml doc. 

I can add a better example to clarify it.

Reply:
@zivkan When I looked yesterday, I concluded that those analyzers were not installed and I didn't want to propose an addition of new analyzers.
There might be *a lot* of things to be suppressed if we were to do that.

Reply:
Visual Studio provides a 3 click way to create suppressions for every instance in the project (maybe even solution?). It's not a high effort task, if that is the concern.

Reply:
I'm not worried about the amount of effort. 
I just don't want our suppression file to be half warnings for comments. 
There's more importance things to tackle imo.

Reply:
Personally, I don't like the global supressions file, because when working on the `cs` file, you have no idea that a warning was suppressed. This makes it more difficult to apply the "boy scout rule" of improving code close to what you're editing already.

But in any case, there is tooling that can help us apply this rule, meaning we don't have to worry about it during code reviews, as per [this comment](https://github.com/NuGet/NuGet.Client/pull/4051#discussion_r635643302). But if we keep the analyzer disabled, then the tooling won't even run, and then the only way this guideline can be enforced is if we do check for it manually when reviewing other people's PRs.  This guidelines document is fairly long already, and it's difficult to keep all the different guidelines in mind when writing code

>  I concluded that those analyzers were not installed

I believe the analyzers are built into Roslyn, not additional packages that need to be referenced. For example [cs1591](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-messages/cs1591) [cs1573](https://docs.microsoft.com/en-us/dotnet/csharp/misc/cs1573), [cs1574](https://docs.microsoft.com/en-us/dotnet/csharp/misc/cs1574), appear to be part of the compiler. The docs certainly don't suggest that they're part of a different package/analyzer.  However, in at least one project, we explicitly `NoWarn` them:

https://github.com/NuGet/NuGet.Client/blob/9500c09af1cb9a79dd9df43775922e467d7ebe2c/src/NuGet.Core/NuGet.Frameworks/NuGet.Frameworks.csproj#L10

Therefore any developer who unintentionally violates the warning will never know.

Reply:
> I believe the analyzers are built into Roslyn, not additional packages that need to be referenced. For example cs1591 cs1573, cs1574, appear to be part of the compiler. T

That explains why I could find them in the analyzers warnings :D 

22 projects suppress `cs1591`, so pretty much all of the ones that generate documentation files. 

I played with enabling some of the warnings (I really removed all suppressed warnings in NuGet.Common) and the analyzers enforce the following. 

* The existence of a summary element on public APIs.
* The existence of all or none param elements on public APIs.

What these analyzers do not enforce:
* Actual content in the summary or param elements.

So as is, those rules are *not as strict* as our guidance is, thus the guidance is actually necessary either way.

@zivkan are you suggesting removing this warning from the project files and suppress it in global suppressions instead?

Reply:
Considering the main feedback we get about the NuGet SDK, the few times we get feedback about the NuGet SDK, is that there's a lack of documentation, I think removing the NoWarns will encourage developers creating public APIs to add docs. It might also encourage developers creating new APIs to think a little more about good API design, which I'd also consider a good thing. If it happens, that'd be a great unintended consequence, in my opinion.

I'm fine any way for the non-SDK projects. If we don't want public API XMLDoc for APIs that customers can never call (in our VS assemblies, for example), I won't complain.

However, since the analyzers don't enforce non-empty XMLDoc, then it will require a team culture of writing meaningful docs if we want to improve the NuGet SDK feedback. Therefore, we'd probably need to actually comment on it in code reviews, contrary to my interpretation of @zkat's other comment.

I'd also prefer inline, rather than global, suppressions. But I know you in particular don't like commits that make small modifications to a large number of files.

Reply:
CS* warnings cannot be suppressed in a global suppressions file.
Only CA ones can be suppressed: 
https://docs.microsoft.com/en-us/visualstudio/code-quality/in-source-suppression-overview?view=vs-2019#suppress-violations-using-a-global-suppression-file

Reply:
> I'm fine any way for the non-SDK projects. If we don't want public API XMLDoc for APIs that customers can never call (in our VS assemblies, for example), I won't complain.

I agree. No need to overdo it from the onset. 

> However, since the analyzers don't enforce non-empty XMLDoc, then it will require a team culture of writing meaningful docs if we want to improve the NuGet SDK feedback. Therefore, we'd probably need to actually comment on it in code reviews, contrary to my interpretation of @zkat's other comment.

I agree with @zivkan. Part of code reviews is actually reading the documentation as well. Docs existent is not our bar, it's accurate docs. 

> I'd also prefer inline, rather than global, suppressions. But I know you in particular don't like commits that make small modifications to a large number of files.

Given that there are ~22 files I'd rather that be handled in a separate work item. 
It'd be great if the API were documented instead of just suppressing them. 
We could also potentially remove some redundant APIs that we discover in that effort. 6.0/ Dev17 is as good as any for that type of change.

Reply:
> It'd be great if the API were documented instead of just suppressing them.

I agree. I just didn't want the burden of documenting the APIs on the person who stops using NoWarn